macrotranslator.py

```diff
@@ -0,0 +1,87 @@
+from macros import MacroMap, Macro, Invocation
+from translationconfig import TranslationConfig
+import logging
+import re
+
+logger = logging.getLogger(__name__)
+
+
+class MacroTranslator:
+    def __init__(self, translation_config: TranslationConfig) -> None:
+        self.translation_config = translation_config
+
+    def generate_macro_translations(self,
+                                    mm: MacroMap) -> dict[Macro, str | None]:
+        translationMap: dict[Macro, str | None] = {}
+
+        for macro, invocations in mm.items():
+            translationMap[macro] = self.get_macro_translation(macro, invocations)
+
+        return translationMap
+
+    def get_macro_translation(self, macro: Macro, invocations: set[Invocation]) -> str | None:
+
+        if self.should_skip_macro(macro, invocations):
+            return None
+
+        if macro.IsFunctionLike:
+            return self.translate_function_like_macro(macro, invocations)
+        elif macro.IsObjectLike:
+            return self.translate_object_like_macro(macro, invocations)
+
+    def should_skip_macro(self, macro: Macro, invocations: set[Invocation]) -> bool:
```

Comment:
Ideally any of the logic for checking if we should translate a macro should be in the `ie_def()` function in `predicates/interface_equivalent.py` so that we keep it all in one place (also makes it easier to write about our translation rules in the paper to have them all in one place as well).

Reply:
You don't have to move this function's logic to the `ie_def()` necessarily; I'm just asking that all the logic for checking if a macro is translatable be in one function/method for concision

Reply:
> Ideally any of the logic for checking if we should translate a macro should be in the `ie_def()` function in `predicates/interface_equivalent.py` so that we keep it all in one place (also makes it easier to write about our translation rules in the paper to have them all in one place as well).

I more so put skips here because I didn't feel that it fit in ie_def (defining what is and isn't interface equivalent) but was more because of limitations with Maki. I.e we could translate macros containing a function pointer, but Maki doesn't properly emit the type signature. And we could translate macros with a DeclRefExpr, but that's a more complex case to handle.

Do you have any more thoughts on structuring this? I wouldn't mind moving it to ie_def but I'd have to restructure to make tracking the amount of skips due to these two reasons possible.

Reply:
Ok I see what you mean; separating the theoretical limitations from the technical limitations.
Feel free to leave these checks here then; just add a comment making it clear that the checks here are implementation details preventing translation and not because of semantic issues that would arise from replacing the macro definition with a variable/enum/function.

> Do you have any more thoughts on structuring this?

If you wanted to restructure then what I would do is print the tracking information for each property first, and then perform the translatability check for `should_skip_macro()`, the perform the regular check for `is_def()`. The new definition of `ie_def()` would look something like this:

```python3
def ie_def(m: Macro, pd: PreprocessorData) -> bool:
    is_ = pd.mm[m]
    # We only analyze top-level non-argument invocations
    assert all([i.IsTopLevelNonArgument for i in is_])

    # First check for conditions that are due to technical limitations of Maki and MerC and not
    # due to irreconcilable differences in macro and C semantics.

    # If any part of the type signature has a function pointer
    invocation_has_function_type = \
        any(i.IsExpansionTypeFunctionType or i.IsAnyArgumentTypeFunctionType for i in invocations)

    # TODO(Joey): Implement a way to translate these
    if invocation_has_function_type:
        logger.debug(f"Skipping {macro.Name} as it has a function pointer type")

    # If body contains a DeclRefExpr and is in a header file, skip
    # TODO(Joey/Brent): Find better way on Maki side to handle this
    invocation = next(iter(invocations))

    invocation_has_decl_ref_expr = invocation.DoesBodyContainDeclRefExpr
    if invocation_has_decl_ref_expr and invocation.DefinitionLocationFilename.endswith(".h"):
        logger.debug(f"Skipping {macro.Name} as it contains a DeclRefExpr")

    if (invocation_has_function_type or
        (invocation_has_decl_ref_expr and invocation.DefinitionLocationFilename.endswith(".h")):
        return False

    # Then check for semantic differences that would make replacing
    # this macro definition wiht C code incorrect.

    # ... Rest of original definition of ie_def()
    # (maybe we add logging statements for these checks as well)
```

Reply:
For now I'm going to rename the function to clarify and consider revisiting our skip structure later (possibly returning a specific reason)